packages/dds/tree/src/shared-tree/transactionTypes.ts

```diff
@@ -0,0 +1,66 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { ImplicitFieldSchema, TreeNode } from "../simple-tree/index.js";
+
+/**
+ * A special object that signifies when a SharedTree {@link RunTransaction | transaction} should "roll back".
+ * @public
+ */
+export const rollback = Symbol("SharedTree Transaction Rollback");
+
+/**
+ * A requirement for a SharedTree transaction to succeed.
+ * @remarks Transaction constraints are useful for validating that the state of the tree meets some requirement when a transaction runs.
+ * In general, when running a transaction a client can validate their tree state in whatever way they wish and decide to either proceed with the transaction or not.
+ * However, they cannot know what the tree state will be when the transaction is _sequenced_.
+ * There may have been any number of edits from other clients that get sequenced before the transaction is eventually sequenced.
+ * Constraints provide a way to validate the tree state after the transaction has been sequenced and abort the transaction if the constraints are not met.
+ * All clients will validate the constraints of a transaction when it is sequenced, so all clients will agree on whether the transaction succeeds or not.
+ * @public
+ */
+export type TransactionConstraint = NodeInDocumentConstraint; // TODO: Add more constraint types here
+
+/**
+ * A transaction {@link TransactionConstraint | constraint} which requires that the given node exists in the tree.
+ * @remarks The node must be in the document (its {@link TreeStatus | status} must be {@link TreeStatus.InDocument | InDocument}) to qualify as "existing".
+ * @public
+ */
+export interface NodeInDocumentConstraint {
+	readonly type: "nodeInDocument";
+	readonly node: TreeNode;
+}
+
+/**
+ * Parameters for running a transaction on the tree view that applies one or more edits to the tree as a single atomic unit.
+ * @alpha
+ */
+export interface TransactionParams<TResult> {
+	/**
+	 * The function to run as the body of the transaction. This function is passed the root of the tree.
+	 * At any point during the transaction, the function may return the special {@link RunTransaction.rollback | rollback value}
+	 * (`Tree.runTransaction.rollback`) to abort the transaction and discard any changes it made so far.
+	 */
+	readonly transaction: () => TransactionStatus<TResult>;
```

Comment:
I wonder if we should make this
```suggestion
	readonly transaction: () => TransactionStatus<TResult> | undefined;
```
so that transaction authors don't have to return anything if they don't want to.

Reply:
I think that's a good idea. I don't see any reason to prevent users from having a "fast path" for simple transactions that don't want to bother with a return value and are always successful (no return => committed/success).

Ideally I think we'd want something like `TransactionResult<TResult> | void`, rather than `undefined`. But, I'm not sure that is correct - because a function can return _anything_ and still satisfy a return value of `void`. So, we'd have to use a type guard to guess whether the return value is a transaction result, which isn't guaranteed to never collide with some other user type that happen to be returned from their drop-in transaction function (fringe and unlikely scenario, I know...).

Following that logic, I think I see three reasonable possibilities:

1. `() => TransactionStatus<TResult> | "committed" | "aborted"`
  If we're going to let them return `undefined` we may as well have them return a status signal instead. However, if they want to pass up a return value then they still need to use the `TransactionStatus`
2. `() => TransactionStatus<TResult> | TResult`
  This makes returning things easier, but allows no shortcut for abort/rollback and has the drawback mentioned above (`TResult` might be some object that looks like a `TransactionStatus` erroneously).

I think I want to get more second opinions on this one from API Council and Nick. It's not clear to me what the best option is. There are other possibilities too - e.g. we could have functions available that make it easy to construct the TransactionStatus object. So:

`return commit()`
`return abort()` (or `return rollback()`)
`return commit(myResult)`
`return abort(myResult)`

But then, where do those functions come from? Are they imported free functions? Or can we hang them off of something relevant that we know the user already has - like we did with the `rollback` symbol?

Reply:
Note that making the return type `TransactionResult<TResult> | undefined` doesn't force them to return anything but it does ensure that they don't return something that isn't a `TransactionResult<TResult>` or `undefined`.